File: javascript/packages/linter/src/rules/erb-no-output-in-attribute-name.ts
import { ParserRule } from "../types.js"
import { BaseRuleVisitor } from "./rule-utils.js"
import { isERBNode, isERBOutputNode } from "@herb-tools/core"

import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { ParseResult, HTMLAttributeNameNode } from "@herb-tools/core"

class ERBNoOutputInAttributeNameVisitor extends BaseRuleVisitor {
  visitHTMLAttributeNameNode(node: HTMLAttributeNameNode): void {
    for (const child of node.children) {
      if (!isERBNode(child)) continue
      if (!isERBOutputNode(child)) continue

      this.addOffense(
        "Avoid ERB output in attribute names. Use static attribute names with dynamic values instead.",
        child.location,
      )
    }

    super.visitHTMLAttributeNameNode(node)
  }
}

export class ERBNoOutputInAttributeNameRule extends ParserRule {
  static ruleName = "erb-no-output-in-attribute-name"

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "error"
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new ERBNoOutputInAttributeNameVisitor(this.ruleName, context)
    visitor.visit(result.value)
    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/erb_no_output_in_attribute_name.rs
use crate::utils::erb_utils::{as_erb_node, is_output_tag_opening};

use herb::nodes::HTMLAttributeNameNode;
use herb::Visitor;

rule_visitor!(ERBNoOutputInAttributeNameVisitor);
define_parser_rule!(
  ERBNoOutputInAttributeNameRule,
  "erb-no-output-in-attribute-name",
  Error,
  ERBNoOutputInAttributeNameVisitor,
  introduced_in: "0.9.0"
);

impl Visitor for ERBNoOutputInAttributeNameVisitor {
  fn visit_html_attribute_name_node(&mut self, node: &HTMLAttributeNameNode) {
    for child in &node.children {
      if let Some((tag_opening, _, location)) = as_erb_node(child) {
        if is_output_tag_opening(tag_opening) {
          self.add_offense(
            "Avoid ERB output in attribute names. Use static attribute names with dynamic values instead.",
            location.clone(),
          );
        }
      }
    }

    self.walk_html_attribute_name_node(node);
  }
}
